capellambse/decl.py

```diff
@@ -0,0 +1,404 @@
+    """A promise could not be fulfilled.
+
+    This exception is raised when a promise is referenced via
+    ``!promise``, but it is never fulfilled by declaring an object with
+    the same ``promise_id``.
+    """
+
+    def __str__(self) -> str:
+        if (
+            len(self.args) == 1
+            and isinstance(self.args[0], cabc.Iterable)
+            and not isinstance(self.args[0], str)
+        ):
+            return ", ".join(i.identifier for i in self.args[0])
+        return super().__str__()
+
+
+@dataclasses.dataclass(frozen=True)
+class UUIDReference:
+    """References a model object by its UUID."""
+
+    uuid: helpers.UUIDString
+
+    def __post_init__(self) -> None:
+        if not helpers.is_uuid_string(self.uuid):
+            raise ValueError(f"Malformed `!uuid`: {self.uuid!r}")
+
+
+class YDMDumper(yaml.SafeDumper):
+    """A YAML dumper with extensions for declarative modelling."""
+
+    def represent_promise(self, data: t.Any) -> yaml.Node:
+        assert isinstance(data, Promise)
+        return self.represent_scalar("!promise", data.identifier)
+
+    def represent_uuidref(self, data: t.Any) -> yaml.Node:
+        assert isinstance(data, UUIDReference)
+        return self.represent_scalar("!uuid", data.uuid)
+
+
+YDMDumper.add_representer(Promise, YDMDumper.represent_promise)
+YDMDumper.add_representer(UUIDReference, YDMDumper.represent_uuidref)
+
+
+class YDMLoader(yaml.SafeLoader):
+    """A YAML loader with extensions for declarative modelling."""
+
+    def construct_promise(self, node: yaml.Node) -> Promise:
+        if not isinstance(node, yaml.ScalarNode):
+            raise TypeError("!promise only accepts scalar nodes")
+        data = self.construct_scalar(node)
+        if not isinstance(data, str):
+            raise TypeError("!promise only accepts string scalars")
+        return Promise(data)
+
+    def construct_uuidref(self, node: yaml.Node) -> UUIDReference:
+        if not isinstance(node, yaml.ScalarNode):
+            raise TypeError("!uuid only accepts scalar nodes")
+        data = self.construct_scalar(node)
+        if not helpers.is_uuid_string(data):
+            raise ValueError(f"Not a well-formed UUID string: {data}")
+        return UUIDReference(data)
+
+
+YDMLoader.add_constructor("!promise", YDMLoader.construct_promise)
+YDMLoader.add_constructor("!uuid", YDMLoader.construct_uuidref)
+
+
+try:
+    import click
+except ImportError:
+
+    def _main() -> None:
+        """Display a dependency error."""
+        print("Error: Please install 'click' and retry", file=sys.stderr)
+        raise SystemExit(1)
+
+else:
+
+    @click.command()
+    @click.option("-m", "--model", type=capellambse.ModelCLI(), required=True)
+    @click.argument("file", type=click.File("r"))
+    def _main(model: capellambse.MelodyModel, file: t.IO[str]) -> None:
+        """Apply a declarative modelling YAML file to a model."""
+        apply(model, file)
+        model.save()
```

Comment:
Is the click CLI meant to be available optionally? This isn't documented and therefore surprises me. I'd suggest to add click as a dependency and get rid of the try/except/else routine.

Reply:
Addressed by adding both - proper documentation *and* the missing dependency. :)